timings/timings.go

```diff
@@ -0,0 +1,106 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timings
+
+import (
+	"fmt"
+	"time"
+)
+
+// Subject is a type of the activity being measured; available subjects are defined below.
+type Subject string
+
+const (
+	// execution of ensure loop
+	Ensure Subject = "ensure"
+	// acitivity related to task
+	Task = "task"
+)
+
+// Timings represents a tree of measures concerning given Subject (activity).
+// Calling Start on the Timings objects returns a root Measure and starts a new performance measurement. Measurement needs to
+// be finished by calling Stop function on the Measure object.
+// Nested measures may be collected by calling StartNested on Measure objects. Similar to the above, nested measurements need
+// to be finished by calling Stop on them.
+//
+// Typical usagage:
+//   timing := timings.New(timings.Ensure)
+//   measure := timing.Start("computation", "...")
+//   ....
+//   nested := measure.StartNested("sub-computation", "...")
+//   ....
+//   nested.Stop()
+//   measure.Stop()
+type Timings struct {
+	subject Subject
+	meta    map[string]string
+
+	// root measurement
+	m *Measure
```

Comment:
I'm coming to the conclusion that it would be easier to make things more symmetric by making this a list too!

Reply:
Done. Note it has a consequence, you can now create Timings object and have sibling timings already on root level, so a duration for the entire tree needs to calculated, which I'm now doing on Save, total duration is calculated from first start time and most recent stop time of any Timing in the tree.